velero-csi-plugin/util.go

```diff
@@ -0,0 +1,141 @@
+	if err != nil {
+		return nil, err
+	}
+
+	for _, p := range podList.Items {
+		for _, v := range p.Spec.Volumes {
+			if v.PersistentVolumeClaim != nil && v.PersistentVolumeClaim.ClaimName == pvcName {
+				podsUsingPVC = append(podsUsingPVC, p)
+			}
+		}
+	}
+
+	return podsUsingPVC, nil
+}
+
+func getPodVolumeNameForPVC(pod corev1api.Pod, pvcName string) (string, error) {
+	for _, v := range pod.Spec.Volumes {
+		if v.PersistentVolumeClaim != nil && v.PersistentVolumeClaim.ClaimName == pvcName {
+			return v.Name, nil
+		}
+	}
+	return "", errors.Errorf("Pod %s/%s does not use PVC %s/%s", pod.Namespace, pod.Name, pod.Namespace, pvcName)
+}
+
+func getPodVolumesUsingRestic(pod corev1api.Pod) []string {
+	resticAnnotation := pod.Annotations[resticPodAnnotation]
+	if resticAnnotation == "" {
+		return []string{}
+	}
+	return strings.Split(pod.Annotations[resticPodAnnotation], ",")
+}
+
+func contains(slice []string, key string) bool {
+	for _, i := range slice {
+		if i == key {
+			return true
+		}
+	}
+	return false
+}
+
+func isPVCBackedUpByRestic(pvcNamespace, pvcName string, podClient corev1client.PodsGetter) (bool, error) {
+	pods, err := getPodsUsingPVC(pvcNamespace, pvcName, podClient)
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	for _, p := range pods {
+		resticVols := getPodVolumesUsingRestic(p)
+		if len(resticVols) > 0 {
+			volName, err := getPodVolumeNameForPVC(p, pvcName)
+			if err != nil {
+				return false, err
+			}
+			if contains(resticVols, volName) {
+				return true, nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+func getVolumeSnapshotClassForStorageClass(provisioner string, snapshotClient snapshotter.SnapshotV1beta1Interface) (*snapshotv1beta1api.VolumeSnapshotClass, error) {
+	snapshotClasses, err := snapshotClient.VolumeSnapshotClasses().List(metav1.ListOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "error listing volumesnapshot classes")
+	}
+	for _, sc := range snapshotClasses.Items {
+		if sc.Driver == provisioner {
+			return &sc, nil
+		}
+	}
+	return nil, errors.Errorf("failed to get volumesnapshotclass for provisioner %s", provisioner)
+}
+
+func getVolumeSnapshotContentForVolumeSnapshot(volSnap *snapshotv1beta1api.VolumeSnapshot, snapshotClient snapshotter.SnapshotV1beta1Interface, log logrus.FieldLogger) (*snapshotv1beta1api.VolumeSnapshotContent, error) {
+	var snapshotContent *snapshotv1beta1api.VolumeSnapshotContent
+	for {
+		vs, err := snapshotClient.VolumeSnapshots(volSnap.Namespace).Get(volSnap.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, fmt.Sprintf("failed to get volumesnapshot %s/%s", volSnap.Namespace, volSnap.Name))
+		}
+
+		// TODO: add timeout
+		if vs.Status == nil || vs.Status.BoundVolumeSnapshotContentName == nil {
+			log.Infof("Waiting for CSI driver to reconcile volumesnapshot %s/%s. Retrying in 5s", volSnap.Namespace, volSnap.Name)
+			time.Sleep(5 * time.Second)
```

Comment:
I don't think we should hold up this PR, but instead of sleeps, this could probably use the `PollImmediate` function, like at https://github.com/vmware-tanzu/velero/blob/master/pkg/restore/restore.go#L636